### DISCUSSION TITLE
suggestions for sharing addons across multiple copilot apps

### DISCUSSION THREAD

**Discussion by codekitchen** [aws/copilot-cli#3130] [Q&A] [answered]
Hi, I'm looking for any suggestions anybody has on a good way to share addon templates between multiple copilot apps, while making it easy for those apps to adopt new versions of the template.

To make the question more concrete, we've created our own RDS addon that uses regular (not serverless v1) Aurora and adds quite a bit of alerting and other goodies. Right now we are just copying that addon template to each copilot service that is using it, but it is already getting very difficult to propagate changes to that template across the different apps, especially given that realistically you'll have to make modifications to at least the parameter defaults to enable different instance sizes, etc. So just copying in a new version of the template isn't enough, you have to do a sort of manual merging process to copy it in while preserving the parameter changes specific to your service.

I took a go at solving this by having the copilot addon be nothing but a CF substack resource (AWS::CloudFormation::Stack) pointing to a shared template in S3. But in the end I realized this won't work currently, because copilot parses the addons template looking for resource names in order to do its magic to find secrets, security groups, etc in the outputs. Since those resources are defined in the substack, copilot can no longer parse them out.

**Comment by codekitchen** [comment]
One solution I've just arrived at is to separate all the tweakable parameters into a separate .yml file, so you have something like this:

```
database-cluster.yml
database-cluster-config.yml
```

So `database-cluster-config.yml` is the only file that individual apps should have to modify, and we can copy over `database-cluster.yml` with a new version when we want to update it.

This definitely feels a bit clunky and will take some training but it is better than my baseline of manually merging CF templates to update. I'm still definitely interested in a better solution.

**Comment by huanjani** [comment, marked as answer]
Hi @codekitchen! 
Does the `svc package` command help at all, for what you're trying to do?
https://aws.github.io/copilot-cli/docs/commands/svc-package/

**Reply by codekitchen** [reply]
Hi @huanjani thank you. I don't think `svc package` helps unless I am missing something -- that command will output the compiled CF templates used to create your service, which I think is mostly useful if you are looking to either debug or eject from copilot and stop using it? I am looking for the best way to create a shared resource template, such as my example of an Aurora Cluster, that I can then easily share across multiple Copilot applications or services within my company.

The main problem I'm seeing is that since in Copilot you modify the CF template directly, changing the `Default:` for any parameter you want to modify, it makes it difficult to share updates to that template between different Copilot apps.

To make it more concrete: let's say we have three applications using Copilot, `app1` `app2` and `app3`. They all need an Aurora DB cluster, and instead of `copilot storage init` they all are going to use our own custom template we've created by copying it into `copilot/<svcname>/addons/ourDB.yml`. But then `app1` wants to change the instance type to say `db.m4.large` while `app2` wants to restore from a snapshot. Now, if we make an improvement to our shared `ourDB.yml` template, each app needs a way to use that improvement while retaining the local changes they've made to things like instance type. This is where I'm looking for a better method to propagate those template changes to the various apps.

Maybe if there was some way to specify in the manifest any overrides for the addon CF template Parameters, something like:

```yaml
addons:
  parameters:
    dbInstanceSize: db.m4.large
```

That would mean we no longer need to modify the addon template directly. But that's just one possibility, maybe there's some better method we could adopt without a new Copilot feature.

**Reply by huanjani** [reply]
Ahhh, gotcha gotcha. Thanks for the clarification. Yeah, `svc package` won't help you with that at all 😂 . 
Configuring the addons template via the manifest is an interesting idea for sure. I'm curious if it's a popular ask.
I can't think of a workaround beyond your solution above, but the Copilot team will noodle on this!

**Reply by codekitchen** [reply]
thanks so much!

**Reply by huanjani** [reply]
My teammate may have the solution for you! 
https://aws.github.io/copilot-cli/docs/developing/additional-aws-resources/#customizing-the-parameters-section

**Reply by codekitchen** [reply]
Ah I hadn't noticed that in the documentation. It sounds like that is the same idea I arrived at in my first response above, but it's good to see it officially documented. I do worry about it being confusing and error-prone for team members who are less familiar with Copilot -- right now we are tackling that as best we can with lots of comments in the yml files to guide people.

**Reply by codekitchen** [reply]
Oh sorry, I misunderstood this documentation until I actually tried it out -- `addons.parameters.yml` is a file with special meaning in Copilot, not just another template name that will be concatenated with the others. It defines actual parameter values that will be passed to CF _as_ parameter values, rather than concatenated into the addon template itself. That is a little bit of an improvement, thank you for pointing me to that.
